packages/expo-updates/ios/EXUpdates/Database/EXUpdatesDatabaseIntegrityCheck.m

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021-present 650 Industries. All rights reserved.
+
+#import <EXUpdates/EXUpdatesAsset.h>
+#import <EXUpdates/EXUpdatesDatabaseIntegrityCheck.h>
+#import <EXUpdates/EXUpdatesFileDownloader.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation EXUpdatesDatabaseIntegrityCheck
+
++ (BOOL)runWithDatabase:(EXUpdatesDatabase *)database
+              directory:(NSURL *)directory
+                 config:(EXUpdatesConfig *)config
+         embeddedUpdate:(nullable EXUpdatesUpdate *)embeddedUpdate
+                  error:(NSError ** _Nullable)error
+{
+  NSError *err;
+  NSArray<EXUpdatesAsset *> *assets = [database allAssetsWithError:&err];
+  if (err) {
+    *error = err;
+    return NO;
+  }
+
+  NSMutableArray<EXUpdatesAsset *> *missingAssets = [NSMutableArray new];
+  dispatch_sync([EXUpdatesFileDownloader assetFilesQueue], ^{
+    for (EXUpdatesAsset *asset in assets) {
+      if (!asset.filename || ![[self class] asset:asset existsInDirectory:directory]) {
```

Comment:
```suggestion
      if (![[self class] asset:asset existsInDirectory:directory]) {
```
Going by your comment, `asset.filename` will always be defined.
Alternatively, we could name a missing asset 'DOES_NOT_EXIST' in lieu of `null`? And then we can check string equality here.

(if I remember correction filenames are created by hashing the file contents and so there is no danger of colliding with something like 'DOES_NOT_EXIST').

Reply:
Nice catch here 😅 

I am hesitant to add a dummy filename here; that would create some really confusing behavior if someone in the future does have an asset with the key `DOES_NOT_EXIST`, however improbable that is. (That also wouldn't address the issue of divergent behavior between iOS and Android.) Would rather write the code in such a way that it doesn't depend on the `filename`/`relativePath` fields to indicate whether or not the file actually exists on disk -- which it already is.

We could modify the Android counterpart of this class to match this PR's behavior, though (not setting `relativePath` to null, since that isn't strictly necessary). I could also follow up with a further PR to add a separate column to SQLite that explicitly indicates whether or not the file exists on disk; though that column would be inherently untrustworthy because the fs can be modified at any time without the database knowing. What do you think?

Reply:
I think it's unideal but fine for the Android and iOS logic to diverge here a little bit as long as the divergence is encapsulated.

> Would rather write the code in such a way that it doesn't depend on the filename/relativePath fields to indicate whether or not the file actually exists on disk

I agree with that.

Reply:
I think what you have here seems good. Modifying the Android counterpart to avoid this check as well seems like the best solution. But it also doesn't seem urgent and time is probably better spend on other stuff.